### PR TITLE
Add support for Julia

### LINKED ISSUE #2: Support for Julia language
Hi,

This seems a great plugin. Could it be implemented for Julia as well? The syntax is very similar to the Ruby one.

Thank you!

### PR DESCRIPTION
**Pull request by jasonrhansen**
Closes #2

There's still one test that isn't passing and I'm not sure what to do for it.

The test that fails is "julia, nested do clause no arg".

```Julia
get(dict, "a") do
  get(dict, "b") do
end
```

This works when the do block has an argument because the second do block is parsed as valid.
image

Here's the tree.
image

But when the do block doesn't have an argument it doesn't work because neither do block is parsed as valid.
image

Here's the tree for the case that doesn't work.
image

Any ideas on how this could be handled?

### CODE REVIEW

**Comment by jasonrhansen** [comment]
Also, I don't use Julia myself so I don't know how complete this is. I was just basing this on the Julia documentation.

File: queries/julia/endwise.scm [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+("module" @indent . (identifier) @cursor (#endwise! "end"))
+("let" @indent (variable_declaration (_) @cursor .) (#endwise! "end"))
+("if" @indent . (binary_expression (_) @cursor .) (#endwise! "end"))
+("for" @indent . (for_binding (_) @cursor .) (#endwise! "end"))
+("while" @indent . (binary_expression (_) @cursor .) (#endwise! "end"))
+("try" @indent @cursor (#endwise! "end"))
+("quote" @indent @cursor (#endwise! "end"))
+("begin" @indent @cursor (#endwise! "end"))
+(("do" @cursor @indent . (identifier)? @cursor . (tuple_expression)? @cursor . (bare_tuple_expression)? @cursor) (#endwise! "end"))
+
```
Comment (RRethy) [inline queries/julia/endwise.scm]:
```suggestion
(("do" @cursor @indent . (bare_tuple_expression)? @cursor) (#endwise! "end"))
```

Comment (RRethy) [reply queries/julia/endwise.scm]: Hmm, this fixes

```julia
get(dict, "a") do
  get(dict, "b") do
end
```

But fails on

```julia
map(1:9) do x
```

Comment (RRethy) [reply queries/julia/endwise.scm]: The julia parser is a bit weird, not as weird as the ruby parser at least tho

**Comment by RRethy** [comment]
Thanks for taking this on, I can fix the query I commented on above, going through https://github.com/tree-sitter/tree-sitter-julia/blob/master/src/node-types.json rn. I'm a bit hesitant with adding Julia since I don't use it and you mentioned you don't use it, I would want @philipGaudreau to try this out and go through the test cases to see if anything is missing. As well, anywhere in the README where `Julia` is written should be replaced with `Julia (community-maintained)` so it's clear that support is best-effort.

Edit: Fixed the query above

File: queries/julia/endwise.scm (hunk 1 above)
Comment (RRethy) [reply queries/julia/endwise.scm]: This works:

```suggestion
("do" @indent @cursor . [(bare_tuple_expression) (identifier) (parenthesized_expression)] @cursor (#endwise! "end"))
("do" @indent @cursor (#endwise! "end"))
```

**Review by RRethy — CHANGES_REQUESTED** [review]
Change to README and query I mentioned above. When @philipGaudreau gives his review on the test cases, then I'll merge.

File: queries/julia/endwise.scm (hunk 1 above)
Comment (jasonrhansen) [reply queries/julia/endwise.scm]: That's great, thanks for your help!

Comment (phgz) [reply queries/julia/endwise.scm]: Yes, this is a bizarre syntax that is part of the language IMO. It is a bit like the `with ... as` in python. You can check this article for more details. https://erik-engheim.medium.com/pythons-with-as-statement-in-julia-javascript-lua-and-other-languages-e2d29941959a

**Comment by jasonrhansen** [comment]
I was looking into this [comment](https://github.com/RRethy/nvim-treesitter-endwise/issues/2#issuecomment-1037366986) by @philipGaudreau on #2 and I can confirm there are issues where sometimes `end`s get added when they shouldn't be.

It think the simplest case is this:
```Julia
begin
  |
end
```
after pressing `enter` becomes:
```Julia
begin
  
  |
end
end
```

It seems like it would be simple to fix, but I haven't been able to figure out a way to express the query without this problem. Here's the query I'm using:

```Scheme
("begin" @indent @cursor (#endwise! "end"))
```
But the tree for this
```Julia
begin
  
end
```
looks like this
```
ERROR [0, 0] - [3, 0]
  "begin" [0, 0] - [0, 5]
  identifier [2, 0] - [2, 3]
  "\n" [2, 3] - [3, 0]
```
so the query still matches because it's still in an error state until one or more expressions are added between the `begin` and `end`,  at which point it becomes a `compound_expression`.

After adding an expression like this
```Julia
begin
  1
end
```
the tree looks like this
```
compound_expression [0, 0] - [2, 3]
  "begin" [0, 0] - [0, 5]
  number [1, 2] - [1, 3]
  "\n" [1, 3] - [2, 0]
  "end" [2, 0] - [2, 3]
"\n" [2, 3] - [3, 0]
```
and is no longer an error. It also doesn't match the "begin" query, so it no longer has the same issue.

Do you have any ideas how to fix this problem?

**Comment by RRethy** [comment]
I got some of them working with some code changes and fancy queries, but tbh I'm going to close this PR. I don't see anyone using Julia (me and you don't use it either) and the original poster of the issue has been scarcely available, the Julia parser seems snowflake-y enough that any solution we come up with is going to have bugs that I don't want to maintain.

**Comment by simonmandlik** [comment]
Hi @RRethy, would you be willing to reopen this?

I'm willing to help with all the implementation and testing required. I have little experience with treesitter queries, however, I'm quite proficient in `julia` itself.

For me, even if the plugin is working, say, only 90% of the time, it would be an immense help for me.

**Comment by RRethy** [comment]
> Hi @RRethy, would you be willing to reopen this?
> 
> I'm willing to help with all the implementation and testing required. I have little experience with treesitter queries, however, I'm quite proficient in `julia` itself.
> 
> For me, even if the plugin is working, say, only 90% of the time, it would be an immense help for me.

I'd prefer a new PR than reopening one.

**Comment by simonmandlik** [comment]
I have started working on a new one.

In the meantime, can I please ask you for help with the following? Why is is not matched?

Code:
```julia
module Foo<cursor>
```

TS Tree:
```
"module" ; [1:1 - 6]
(identifier) ; [1:8 - 10]
"\n" ; [1:11 - 2:0]
```

Query:
```
("module" @indent . (identifier) @cursor (#endwise! "end"))
```

**Comment by RRethy** [comment]
I'm not sure 😕 that seems like it should work. Maybe try seeing if the nvim-treesitter devs would know.

**Pull request closed without merging**